mart/callbacks/adversary_connector.py

```diff
@@ -7,15 +7,51 @@
 from __future__ import annotations
 
 import types
-from typing import Callable
+from typing import Any, Callable
 
+import torch
 from lightning.pytorch.callbacks import Callback
 
-from ..utils import MonkeyPatch
+from ..utils import MonkeyPatch, get_pylogger
+
+logger = get_pylogger(__name__)
+
 
 __all__ = ["AdversaryConnector"]
 
 
+class training_mode:
+    """A context that switches a torch.nn.Module object to the training mode except for some
+    children."""
+
+    def __init__(self, module, excludes=[]):
+        self.module = module
+        self.excludes = excludes
+
+    def __enter__(self):
+        # Save the original training mode status.
+        self.training = self.module.training
+        self.module.train(True)
+        # Set some children modules of "excludes" to eval mode instead.
+        self.selective_eval_mode("", self.module, self.excludes)
```

Comment:
What is going on with the empty string?

Reply:
We don't know the variable name of the model, so the module path starts with a dot. This is for debug logging only, which print messages like this 

```
Set .model.student_model.feature_extractor.layer3[1].bn1: BatchNorm2d to eval mode.
```